internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1072,77 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
 	}
 	return nil, nil
 }
+
+// checkOldSnapshotsPipelineRun sorts all snapshots for application and cancells all running interationTest pipelineruns within application
+// removes finalizer before the pipelinerun is set as cancelled run finally to be gracefuly cancelled
+func (a *Adapter) checkAndCancelOldSnapshotsPipelineRun(application *applicationapiv1alpha1.Application, snapshot *applicationapiv1alpha1.Snapshot) error {
+
+	snapshots, err := a.loader.GetAllSnapshotsForPR(a.context, a.client, application, snapshot.GetLabels()[gitops.PipelineAsCodePullRequestAnnotation], snapshot.GetLabels()[gitops.PipelineAsCodeRepoURLAnnotation])
+	if err != nil {
+		a.logger.Error(err, "Failed to fetch Snapshots for the application",
+			"application.Name:", application.Name)
+		return err
+	}
+	sortedSnapshots := gitops.SortSnapshots(*snapshots)
+	if len(sortedSnapshots) < 1 {
+		a.logger.Info("There are no found snapshots for given repo at the moment.")
+		return nil
+	}
+	// check for older snapshots, their current status and type
+	for i := 1; i < len(sortedSnapshots); i++ {
+		if gitops.IsComponentSnapshot(&sortedSnapshots[i]) {
+			err = gitops.MarkSnapshotAsCanceled(a.context, a.client, &sortedSnapshots[i], "Snapshot canceled/superseded")
+			if err != nil {
+				a.logger.Error(err, "Failed to mark snapshot as canceled", "snapshot.Name", snapshot.Name)
+				return err
+			}
+			// get all integration pipelineruns for a snapshot
+			err = a.cancelAllPipelineRunsForSnapshot(&sortedSnapshots[i])
+
+		}
+	}
+	return err
+}
+
+func (a *Adapter) cancelAllPipelineRunsForSnapshot(snapshot *applicationapiv1alpha1.Snapshot) error {
```

Comment:
Can we name this function in a way that shows we are only canceling all except the latest PLR? Also, can we please have a header comment for this function to briefly describe what it's doing?

Reply:
Hmm, we're calling this function for older snapshots, so do we really need to not cancel the latest PLR?

Reply:
Thanks for updating. I'm still curious about why we are calling this function on an older snapshot because it now cancels all but the latest Intg PLR for that older snapshot. Do we want that? Or do we want to cancel **ALL** the Intg PLRs for the older snapshots?

Reply:
thanks for the update!